### PR TITLE
[Console] Ease writing to stderr using SymfonyStyle

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #15986
| License       | MIT
| Doc PR        | n/a

### CODE REVIEW

**Comment by fabpot** [comment]
Can't be in 2.8, should be in master

**Comment by chalasr** [comment]
Changed the scope then, proposing `SymfonyStyle::getErrorIo()` to ease writing to stderr using styles (feature), and secondly made built-in commands using it when it's relevant. Let me know if it should be done in two PRs.

**Comment by nicolas-grekas** [comment]
@chalasr you decide :)

**Comment by stof** [comment]
Even adding a new method must be in master, as it is a new feature

File: src/Symfony/Bundle/FrameworkBundle/Command/ConfigDebugCommand.php [hunk 1]
```diff
@@ -58,13 +58,12 @@ protected function configure()
...
 
         if (null === $name = $input->getArgument('name')) {
-            $this->listBundles($io);
-            $io->comment('Provide the name of a bundle as the first argument of this command to dump its configuration. (e.g. <comment>debug:config FrameworkBundle</comment>)');
-            $io->comment('For dumping a specific option, add its path as the second argument of this command. (e.g. <comment>debug:config FrameworkBundle serializer</comment> to dump the <comment>framework.serializer</comment> configuration)');
-
-            return;
+            $this->listBundles($errorIo);
+            $errorIo->comment('Provide the name of a bundle as the first argument of this command to dump its configuration. (e.g. <comment>debug:config FrameworkBundle</comment>)');
+            $errorIo->comment('For dumping a specific option, add its path as the second argument of this command. (e.g. <comment>debug:config FrameworkBundle serializer</comment> to dump the <comment>framework.serializer</comment> configuration)');
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/Command/ConfigDebugCommand.php]: removing the return is broken

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by chalasr** [comment]
@stof It was about proposing the new method here and fixing the existing commands in another PR by creating the instance of `SymfonyStyle` in the commands themselves. see #20632

File: src/Symfony/Bundle/FrameworkBundle/Command/ConfigDebugCommand.php (hunk 1 above)
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Command/ConfigDebugCommand.php]: 👍  fixed in #20632

File: src/Symfony/Component/Console/Tests/Style/SymfonyStyleTest.php [hunk 2]
```diff
@@ -71,4 +72,29 @@ public function inputCommandToOutputFilesProvider()
 
         return array_map(null, glob($baseDir.'/command/command_*.php'), glob($baseDir.'/output/output_*.txt'));
     }
+
+    public function testGetErrorIo()
+    {
+        $input = $this->getMock('Symfony\Component\Console\Input\InputInterface');
```
Comment (chalasr) [inline src/Symfony/Component/Console/Tests/Style/SymfonyStyleTest.php]: thanks, fixed

File: src/Symfony/Component/Console/Style/SymfonyStyle.php [hunk 3]
```diff
@@ -338,6 +338,21 @@ public function newLine($count = 1)
     }
 
     /**
+     * Returns a new instance which makes use of stderr if available, the current instance
+     * otherwise.
+     *
+     * @return self
+     */
+    public function getErrorIo()
```
Comment (fabpot) [inline src/Symfony/Component/Console/Style/SymfonyStyle.php]: The method name looks weird to me as we don't use IO anywhere in the code IIRC

Comment (chalasr) [inline src/Symfony/Component/Console/Style/SymfonyStyle.php]: I agree. `getErrorStyle()` ?

Comment (chalasr) [inline src/Symfony/Component/Console/Style/SymfonyStyle.php]: Changed to `getErrorStyle()`

File: src/Symfony/Bundle/FrameworkBundle/Command/ConfigDebugCommand.php (hunk 1 above)
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Command/ConfigDebugCommand.php]: Could you please dismiss your review? As the change is no more in this PR

File: src/Symfony/Component/Console/Style/OutputStyle.php [hunk 4]
```diff
@@ -145,4 +146,13 @@ public function isDebug()
     {
         return $this->output->isDebug();
     }
+
+    protected function getErrorOutput()
+    {
+        if (!$this->output instanceof ConsoleOutputInterface) {
+            return;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Console/Style/OutputStyle.php]: should it throw instead?

File: src/Symfony/Component/Console/Style/OutputStyle.php [hunk 5]
```diff
@@ -145,4 +147,13 @@ public function isDebug()
     {
         return $this->output->isDebug();
     }
+
+    protected function getErrorOutput()
+    {
+        if (!$this->output instanceof ConsoleOutputInterface) {
+            throw new LogicException(sprintf('The output must be an instance of %s for using getErrorOutput().', ConsoleOutputInterface::class));
```
Comment (fabpot) [inline src/Symfony/Component/Console/Style/OutputStyle.php]: What about returning the regular output instead?

Comment (chalasr) [inline src/Symfony/Component/Console/Style/OutputStyle.php]: Agreed, updated

**Comment by fabpot** [comment]
Thank you @chalasr.

**Pull request merged**